### ISSUE TITLE
problem with :ONCONFIRM call back for fields [BUG]

### ISSUE THREAD

**Issue by nuggreat** [KSP-KOS/KOS#2868]
Having buttons with an `:ONCLICK` call back can trigger the `:ONCHANGE` suffix of a textField apparently depending on the order the widgets are added to the GUI structure.

In the code below clicking on the "up" button will trigger the `:ONCONFIRM` call back of the field many times depending on what the user does.  The initial click will trigger an `:ONCONFIRM` call, each movement of the mouse while the click is held even moving off the button will trigger a call (likely once for every frame of movement), and the release of the button will trigger a call though only if moused over the button.

Where as trying the same on the "down" button will not produce any unexpected `:ONCONFIRM` callbacks.

```
CLEARGUIS().
LOCAL int IS GUI(200).
 LOCAL ilButtonUp IS int:ADDBUTTON("up").
 LOCAL ilField IS int:ADDTEXTFIELD("string").
 LOCAL ilButtonDown IS int:ADDBUTTON("down").
  SET ilButtonDown:ONCLICK TO { PRINT "down button". }.
  SET ilButtonUp:ONCLICK TO { PRINT "up button". }.
  SET ilField:ONCONFIRM  TO { PARAMETER thing. PRINT "field was confirmed". }.
int:SHOW.

BRAKES OFF.
WAIT UNTIL BRAKES.
CLEARGUIS().
```

In my testing the order in which the callbacks where added doesn't appear to have much affect to the operation of this bug beyond delaying it until after the field's call back has gone off once.

I also tested with more buttons and it appears that only buttons added before the field will cause the issue and buttons added after will not.

I have not tested if nesting the buttons and field into there own BOX/LAYOUT will "fix" the issue.

This might be related to issue #2708 though I am not positive on that as the symptoms appear different.

**Comment by Dunbaratu** [comment]
@nuggereat - the problem description doesn't match the example.  The description talks about a "left" and "right" button, when the example makes an "up" and "down" button.  I was confused a while thinking you were talking about the mouse buttons being left or right.

**Comment by nuggreat** [comment]
Whoops sorry about that, I originally started with a left and right but simplified the case (removing a horizontal layout) and forgot to change the description.  I have edited to to clean up the description left should have been up and right should have been down.

**Comment by Dunbaratu** [comment]
Seems to be even more wrong - it's not just that clicking the up button makes many ONCONFIRMs happen.  It's also that clicking the up button makes zero of them appear.

I'm collecting together the test results for this and issue #2708 together so I can look at them in one place:


```
Effect when editing the text field then leaving the field.

When confirming by        You get this many
doing this                ONCOFNIRMs
=======================   =================

Press Enter in the text
field.                      1. (right)

Click outside the GUI
entirely, on the game       0. (wrong)
background.

Click the "down" button.    1. (right)

Click the "up" button.      >1. (wrong)
```

**Comment by nuggreat** [comment]
it isn't clicking out side of the GUI it is having clicking on the `UP` button and then moving the mouse with out releasing the click

**Comment by Dunbaratu** [comment]
I wrote the results down wrong in the table.  Things were shifted.  Look again.

**Comment by Dunbaratu** [comment]
Confirmed that the order of the buttons is what matters, not the order of the hooks.  (Swapping the order of these three lines changes the result:
```
 LOCAL ilButtonUp IS int:ADDBUTTON("up").
 LOCAL ilField IS int:ADDTEXTFIELD("string").
 LOCAL ilButtonDown IS int:ADDBUTTON("down").
```
Any button that appears before the text field has the "too many confirms" bug, and any button that appears after the text field has the "zero confirms" bug.

**Comment by nuggreat** [comment]
There shouldn't be any confirms from a field when you press a button unless the button is interacting with the field in some way perhaps setting the `:TEXT` so buttons after are not bugged but the buttons before are.

**Comment by Dunbaratu** [comment]
It's not "clicking the button should fire ONCONFIRM." 
It's "text field losing focus after having been edited should fire ONCONFIRM."
Clicking the button should make the focus leave the text field, and it's that reason that I was saying it should fire one confirm (but not many).

**Comment by nuggreat** [comment]
Ah I see, I was only testing the buttons and if they cause the field to fire not cases where the field should but wouldn't

**Comment by Dunbaratu** [comment]
You can see an example of the problem I was showing this way:

1. Edit the text field -  Note the color of the field changes to grey indicating it is 'dirty' (in need of an ONCONFIRM being fired).
2. Click outside the gui window.  Note the color of the field changes to yellow indicating it isn't dirty.  This should coincide with one ONCOFIRM firing off, and it does.
3. Edit the text field again - it goes grey again as it should.
4. Now click the "Down" button.  The text field should be going yellow just like it did when you clicked outside the window.  But it isn't.  And it doesn't fire the ONCONFIRM.

**Comment by Dunbaratu** [comment]
I believe the problem hinges on the way the uiID is assigned for a text field, like so:
```
uiID = GUIUtility.GetControlID(FocusType.Passive) + 1; // Dirty kludge.
```
This uiID field is used later to detect whether the widget that currently has the focus is the one this code is running, like so:
```
            if (GUIUtility.keyboardControl == uiID)
            {
                if (Event.current.keyCode == KeyCode.Return || Event.current.keyCode == KeyCode.KeypadEnter)
                    shouldConfirm = true;
                hadFocus = true;
            }
            else
            {
                if (hadFocus)
                    shouldConfirm = true;
                hadFocus = false;
            }
```
The logic is essentially this "If the widget that Unity thinks has the keyboard focus is me, set hadFocus to true.  Otherwise if another widget has focus, or NO widgets have focus (see below), set it to false.  When this value changes from true to false, that means I just lost focus and should fire ONCONFIRM"

The problem is the combination of these two things: "No widgets have focus" is represented by Unity claiming the control ID that has focus is zero.  But the kOS code is falsely assigning the ``uiID`` to zero some of the time, so it ends up thinking the "no focus" widget (Id=0) is itself, generating all kinds of faulty behavior.

The technique kOS is using to query Unity and ask it "hey, what's my control ID" seems all kinds of wrong to me.  I didn't originally write this code so I have to sort of "read minds" to work out what the intent was, but what it actually does doesn't seem right.  As far as I can tell it seems to be *assuming* that the current text widget will have an ID equal to the lastmost non-text widget plus one, and that makes no sense to me, and definitely could be order dependent.

**Comment by Dunbaratu** [comment]
I'm leaving this here for me to look at tomorrow, as I found it in my google searching but don't have time to implement it tonight:

Someone was asking how you use ``GUIUtility.keyboardControl`` when it returns an ID that you cannot reliably tie to a specific control widget because ``GUIUtility.GetControlID`` is so flaky.

That led to this answer, which I think is what I have to change the kOS code to use:

> The intended way to identify a certain control is to use GUI.SetNextControlName and then use GUI.GetNameOfFocusedControl to determine which control has the focus. You can also use GUI.FocusControl to focus a named control.

So when the textarea is being drawn, SetNextControlName lets me specify what I want to call it, and then GetNameOfFocusedControl can use used more reliably in a way where I get to know for sure it's telling me things in terms that mean something to me.

**Comment by Dunbaratu** [comment]
Arrg, so I still don't know why this happens automatically some times and not other times,  But it seems you can make a button manually take control away from other widgets (like text fields) when clicked by having it call ``GUI.FocusControl()`` using the name you provided by ``GUI.SetNextControlName()`` for itself.

So I can force the expected behavior by naming ALL widgets with ``SetNextControlName()`` in their DoGUI(), and then making all widgets take focus when interacted with by having them call ``GUI.FocusControl()``.

That doesn't fix the double-confirm but I feel it's a necessary first step to fixing it because until the IMGUI widgets *themselves* start behaving consistently I can't figure out how to make kOS's hooks on top of them behave consistently.

**Comment by Dunbaratu** [comment]
Well I have some progress and some confusion.  I have gotten it so all the buttons will pull focus away from the text field when clicked (thus generating an OnCONFIRM) whether they're before or after the text field.

But I still don't know why buttons that occur before the text field spew multiple ONCONFIRMS while ones that come after (now) only cause one.  I'm still working on that inconsistency.

**Comment by Dunbaratu** [comment]
I believe I have a fixed version, in PR #2872 but I will want people who make heavy use of GUIs to try it out and confirm it didn't break things before I merge it into develop.
I may make a special ZIP for such a test.
